### ISSUE TITLE
Implement loss function

### ISSUE THREAD

**Issue by stsievert** [lmcinnes/umap#133]
In [`optimize_layout`][1], some function is minimized with SGD, though it's never instantiated. It would be nice to obtain that loss to know which embedding UMAP considers to be best.

[1]:https://github.com/lmcinnes/umap/blob/53c11a29224fcdd8ab89cadbeacc2f0de9a9ace4/umap/umap_.py#L683

I'd like something like

``` python
from umap import UMAP
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split

data = load_digits().data
data_train, data_test = train_test_split(data)
trans = umap.UMAP().fit(data_train)

trans.loss(data_test)
```

This could help perform tune hyperparameters for a particular dataset. That's already possible in the supervised learning case, but having a loss function available could help in the unsupervised case.

**Comment by stsievert** [comment]
The tSNE loss is in https://github.com/lmcinnes/umap/issues/35#issuecomment-355311005. IIRC, the UMAP loss is the same function with another term.

**Comment by lmcinnes** [comment]
I would have to think about the best way to implement that. The actual loss
function is never fully evaluated during training because it is very
computationally expensive to do so. There might be some options for how to
do this for a small training set; there may also be some reasonable options
for approximating it. I'll have to think for bit about this one -- I agree
that it would be valuable, it's just not easy.

**Comment by stsievert** [comment]
Thanks for the quick response @lmcinnes! I'd be okay with a simple approach at first that would work best with small datasets.

I think in the mean time I'll be using classification accuracy and k-nearest neighbors ([docs][doc]) in the meantime to judge embedding quality. I think I'll choose k-nearest neighbors to minimize the effect of model performance (from UMAP embedding → labels).

[doc]:https://umap-learn.readthedocs.io/en/latest/transform.html

**Comment by stsievert** [comment]
Thanks for thinking about this @lmcinnes. I can see myself playing with the MNIST classification example in the mean time.
